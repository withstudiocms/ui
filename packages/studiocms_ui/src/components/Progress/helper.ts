class ProgressHelper {
	private bar: HTMLDivElement;
	private progress: HTMLElement;

	private value: number;
	private max: number;

	constructor(id: string) {
		this.bar = document.getElementById(id) as HTMLDivElement;
		this.progress = this.bar.firstElementChild as HTMLElement;

		this.value = this.getValue();
		this.max = this.getMax();
	}

	getValue() {
		return Number.parseInt(this.bar.dataset.value as string, 10);
	}

	setValue(value: number) {
		const max = Number.parseInt(this.bar.dataset.max as string, 10);
		const percent = Math.round((value / max) * 100);
		this.progress.style.width = `${percent}%`;
	}

	getMax() {
		return Number.parseInt(this.bar.dataset.max as string, 10);
	}

	setMax(value: number) {
		this.bar.dataset.max = value.toString();
	}

	getPercentage() {
		return Math.round((this.value / this.max) * 100);
	}
}

export { ProgressHelper };
