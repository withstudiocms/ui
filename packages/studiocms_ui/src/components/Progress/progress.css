.sui-progress {
  width: 100%;
  height: 1rem;
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius-sm);
  overflow: hidden;
}

.sui-progress-slider {
  height: 100%;
  width: 0;
  transition: all .75s ease;
  border-radius: var(--radius-sm);
  background-color: hsl(var(--primary-base));
}

.sui-progress.success .sui-progress-slider {
  background-color: hsl(var(--success-base));
}

.sui-progress.warning .sui-progress-slider {
  background-color: hsl(var(--warning-base));
}

.sui-progress.danger .sui-progress-slider {
  background-color: hsl(var(--danger-base));
}

.sui-progress.info .sui-progress-slider {
  background-color: hsl(var(--info-base));
}

.sui-progress.monochrome .sui-progress-slider {
  background-color: hsl(var(--mono-base));
}
