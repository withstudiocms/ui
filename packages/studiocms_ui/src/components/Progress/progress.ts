function loadProgressBars() {
	const allBars = document.querySelectorAll<HTMLDivElement>('.sui-progress');

	for (const bar of allBars) {
		const value = bar.dataset.value;
		const max = bar.dataset.max;
	
		const progress = bar.firstElementChild as HTMLElement;
	
		if (value && max) {
			const percent = Math.round((Number.parseInt(value, 10) / Number.parseInt(max, 10)) * 100);
			progress.style.width = `${percent}%`;
		}
	}
}

document.addEventListener('astro:page-load', loadProgressBars);
