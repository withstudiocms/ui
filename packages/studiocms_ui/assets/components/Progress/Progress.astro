---
import type { StudioCMSColorway } from '../../utils/colors.js';
import './progress.css';
import type { HTMLAttributes } from 'astro/types';

type Props = {
	id: string;
	value?: number;
	max?: number;
	color?: Omit<StudioCMSColorway, 'default'>;
} & HTMLAttributes<'div'>;

const { id, value = 0, max = 100, color = 'primary' } = Astro.props;
---

<div class="sui-progress" class:list={[color]} role="progressbar" id={id} data-max={max} data-value={value}>
  <div class="sui-progress-slider" />
</div>
<script>
  import 'studiocms:ui/scripts/progress';
</script>
