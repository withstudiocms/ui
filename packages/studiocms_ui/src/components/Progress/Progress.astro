---
import type { StudioCMSColorway } from '../../utils/colors.js';
import './progress.css';
import type { HTMLAttributes } from 'astro/types';

type Props = {
	id: string;
	label: string;
	value?: number;
	max?: number;
	color?: Omit<StudioCMSColorway, 'default'>;
} & HTMLAttributes<'div'>;

const { id, label, value = 0, max = 100, color = 'primary' } = Astro.props;
---

<div class="sui-progress" class:list={[color]} role="progressbar" aria-label={label} id={id} data-max={max} data-value={value}>
  <div class="sui-progress-slider" />
</div>
<script>
  import 'studiocms:ui/scripts/progress';
</script>
