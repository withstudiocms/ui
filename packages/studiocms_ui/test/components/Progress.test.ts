import { describe, expect } from 'vitest';
import Progress from '../../src/components/Progress/Progress.astro';
import { test } from '../fixtures/vitest/AstroContainer';

describe('Progress Component', () => {
	test('renders Progress component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Progress, 'Progress', {
			props: { id: 'progress-test', value: 50, max: 100, label: "Progress Bar Fixture" },
		});
		expect(result).toMatchSnapshot();
	});
});
