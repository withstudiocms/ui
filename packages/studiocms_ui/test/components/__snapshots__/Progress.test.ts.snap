// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Progress Component > renders Progress component correctly 1`] = `"<div class="sui-progress primary" role="progressbar" aria-label="Progress Bar Fixture" id="progress-test" data-max="100" data-value="50"> <div class="sui-progress-slider"></div> </div> <script type="module" src="/mock/path/Progress.astro?astro&type=script&index=0&lang.ts"></script>"`;
